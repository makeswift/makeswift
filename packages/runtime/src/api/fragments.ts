import { DocumentNode } from 'graphql'
import { APIResourceType } from './types'
import * as FragmentDocumentNodes from './graphql/generated/fragment-document-nodes'

export const Fragments: Record<APIResourceType, DocumentNode> = FragmentDocumentNodes
