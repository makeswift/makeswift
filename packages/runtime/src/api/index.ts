export * from './types'
export * from './fragments'
