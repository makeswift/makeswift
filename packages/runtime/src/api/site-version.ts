import { z } from 'zod'

export const makeswiftSiteVersionSchema = z.enum(['Live', 'Working'])
export const MakeswiftSiteVersion = makeswiftSiteVersionSchema.Enum
export type MakeswiftSiteVersion = z.infer<typeof makeswiftSiteVersionSchema>

export const API_HANDLER_SITE_VERSION_HEADER = 'X-Makeswift-Site-Version'
