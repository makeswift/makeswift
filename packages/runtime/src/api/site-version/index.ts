import { z } from 'zod'
import { jwtDecode } from 'jwt-decode'

export const ApiHandlerHeaders = {
  SiteVersion: 'X-Makeswift-Site-Version',
}

const decodedTokenPayloadSchema = z.object({
  exp: z.number(),
})

type DecodedTokenPayload = z.infer<typeof decodedTokenPayloadSchema>

export const siteVersionSchema = z.object({
  version: z.string(),
  token: z.string(),
})

export type SiteVersion = z.infer<typeof siteVersionSchema>

function millisecondsToSeconds(ms: number): number {
  return Math.floor(ms / 1000)
}

export function serializeSiteVersion(data: SiteVersion): string {
  const jsonString = JSON.stringify({ version: data.version, token: data.token })
  return jsonString
}

function parseSerializedSiteVersion(input: string): SiteVersion | null {
  try {
    return siteVersionSchema.parse(JSON.parse(input))
  } catch {
    console.error('Failed to parse serialized site version:', input)
    return null
  }
}

function getSiteVersionTokenPayload(siteVersion: SiteVersion): DecodedTokenPayload {
  try {
    const decoded = jwtDecode<DecodedTokenPayload>(siteVersion.token)
    return decodedTokenPayloadSchema.parse(decoded)
  } catch (error) {
    console.error('Invalid site version token')
    throw error
  }
}

function secondsUntilTokenExpiration(tokenPayload: DecodedTokenPayload): number {
  const currentTime = millisecondsToSeconds(Date.now())
  return Math.max(tokenPayload.exp - currentTime, 0)
}

export function secondsUntilSiteVersionExpiration(siteVersion: SiteVersion): number {
  try {
    const tokenPayload = getSiteVersionTokenPayload(siteVersion)
    return secondsUntilTokenExpiration(tokenPayload)
  } catch (error) {
    console.error('Failed to check site version expiration:', error)
    return 0
  }
}

function isSiteVersionExpired(siteVersion: SiteVersion): boolean {
  return secondsUntilSiteVersionExpiration(siteVersion) <= 0
}

export function deserializeSiteVersion(input: string): SiteVersion | null {
  try {
    const parsedSiteVersion = parseSerializedSiteVersion(input)
    if (parsedSiteVersion == null) return null

    if (isSiteVersionExpired(parsedSiteVersion)) {
      console.error('Site version token is expired')
      return null
    }
    return parsedSiteVersion
  } catch {
    console.error('Failed to deserialize site version:', input)
    return null
  }
}
