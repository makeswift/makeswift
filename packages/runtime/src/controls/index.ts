export * from './control'

export {
  ControlDefinition,
  Checkbox,
  CheckboxDefinition,
  Color,
  ColorDefinition,
  Combobox,
  ComboboxDefinition,
  Font,
  FontDefinition,
  Group,
  GroupDefinition,
  GroupControl,
  IconRadioGroupDefinition,
  type IconRadioGroupIcon,
  unstable_IconRadioGroup,
  Image,
  ImageDefinition,
  List,
  ListDefinition,
  ListControl,
  Number,
  NumberDefinition,
  Select,
  SelectDefinition,
  Shape,
  ShapeDefinition,
  ShapeControl,
  Style,
  StyleDefinition,
  type StyleProperty,
  StyleControl,
  TextInput,
  TextInputDefinition,
  TextArea,
  TextAreaDefinition,
  unstable_Typography,
  unstable_TypographyDefinition,
} from '@makeswift/controls'

export { Link, LinkDefinition } from './link'
export { RichTextV1Definition, RichTextV1Control } from './rich-text'
export { RichText, RichTextV2Definition, RichTextV2Control } from './rich-text-v2'
export { Slot, SlotDefinition, SlotControl } from './slot'
export { unstable_StyleV2, StyleV2Definition, StyleV2Control } from './style-v2/style-v2'
