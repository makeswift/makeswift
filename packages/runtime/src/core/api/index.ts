/**
 * Re-export all API-related types and utilities
 */
export * from './client';
export * from './types';