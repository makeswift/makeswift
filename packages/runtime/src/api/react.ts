import { type FetchableValue } from '@makeswift/controls'
import * as MakeswiftApiClient from '../state/makeswift-api-client'
import {
  APIResourceType,
  File,
  GlobalElement,
  LocalizedGlobalElement,
  Page,
  PagePathnameSlice,
  Site,
  Snippet,
  Swatch,
  Table,
  Typography,
} from './types'
import { GraphQLClient } from './graphql/client'
import { CreateTableRecordMutation } from './graphql/documents'
import {
  CreateTableRecordMutationResult,
  CreateTableRecordMutationVariables,
} from './graphql/generated/types'

export type SerializedLocalizedResourcesMap = {
  [resourceId: string]: string | null
}

type LocalizedResourcesMap = Map<string, string | null>

export type CacheData = MakeswiftApiClient.SerializedState

export type MakeswiftClientOptions = {
  uri: string
  cacheData?: CacheData
  localizedResourcesMap?: SerializedLocalizedResourcesMap
  locale?: Intl.Locale
}

/**
 * NOTE(miguel): This "client" is used to fetch Makeswift API resources needed for the host. For
 * example, swatches, files, typographies, etc. Ideally it's internal to the runtime and is only
 * used by controls to transform API references to API resources.
 *
 * Moreover, its use should be reserved for the builder only, since for live pages all Makeswift
 * API resources should be embedded in the "page snapshot". In the builder, this client serves the
 * purpose of sending requests for API resources and keeping a cache so that changes that happen in
 * the builder, like modifying a swatch, can be sent via `postMessage` to the host and the cache can
 * immediately update the value and re-render.
 *
 * Furthermore, the API resources requested shouldn't be requested directly from the Makeswift API
 * as that would require those resources to not be authenticated since the requests come from the
 * browser when running the host. Instead, the requests should go to the host directly, at the
 * Makeswift API endpoint (i.g., `/api/makeswift/[...makeswift]` dynamic route) where the host's
 * API key can be used, securely, in the server. For this reason, this client should really be a
 * client of the host's API, not Makeswift's, intended to build and continuously maintain a realtime
 * snapshot for use in the builder, not the lives pages.
 */
export class MakeswiftHostApiClient {
  graphqlClient: GraphQLClient
  makeswiftApiClient: MakeswiftApiClient.Store
  subscribe: MakeswiftApiClient.Store['subscribe']
  private localizedResourcesMap: LocalizedResourcesMap
  private locale: Intl.Locale | undefined

  constructor({ uri, cacheData, localizedResourcesMap = {}, locale }: MakeswiftClientOptions) {
    this.graphqlClient = new GraphQLClient(uri)
    this.makeswiftApiClient = MakeswiftApiClient.configureStore({ serializedState: cacheData })
    this.subscribe = this.makeswiftApiClient.subscribe
    this.localizedResourcesMap = new Map(Object.entries(localizedResourcesMap))
    this.locale = locale
  }

  readSwatch(swatchId: string): Swatch | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.Swatch,
      swatchId,
    )
  }

  async fetchSwatch(swatchId: string): Promise<Swatch | null> {
    return await this.makeswiftApiClient.dispatch(
      MakeswiftApiClient.fetchAPIResource(APIResourceType.Swatch, swatchId),
    )
  }

  resolveSwatch(swatchId: string | undefined): FetchableValue<Swatch | null> {
    return this.resolveResource(APIResourceType.Swatch, {
      id: swatchId,
      read: id => this.readSwatch(id),
      fetch: id => this.fetchSwatch(id),
    })
  }

  readFile(fileId: string): File | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.File,
      fileId,
    )
  }

  async fetchFile(fileId: string): Promise<File | null> {
    return await this.makeswiftApiClient.dispatch(
      MakeswiftApiClient.fetchAPIResource(APIResourceType.File, fileId),
    )
  }

  resolveFile(fileId: string | undefined): FetchableValue<File | null> {
    return this.resolveResource(APIResourceType.File, {
      id: fileId,
      read: id => this.readFile(id),
      fetch: id => this.fetchFile(id),
    })
  }

  readTypography(typographyId: string): Typography | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.Typography,
      typographyId,
    )
  }

  async fetchTypography(typographyId: string): Promise<Typography | null> {
    return await this.makeswiftApiClient.dispatch(
      MakeswiftApiClient.fetchAPIResource(APIResourceType.Typography, typographyId),
    )
  }

  resolveTypography(typographyId: string | undefined): FetchableValue<Typography | null> {
    return this.resolveResource(APIResourceType.Typography, {
      id: typographyId,
      read: id => this.readTypography(id),
      fetch: id => this.fetchTypography(id),
    })
  }

  readGlobalElement(globalElementId: string): GlobalElement | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.GlobalElement,
      globalElementId,
    )
  }

  async fetchGlobalElement(globalElementId: string): Promise<GlobalElement | null> {
    return await this.makeswiftApiClient.dispatch(
      MakeswiftApiClient.fetchAPIResource(APIResourceType.GlobalElement, globalElementId),
    )
  }

  readLocalizedGlobalElement(globalElementId: string): LocalizedGlobalElement | null {
    const localizedGlobalElementId = this.getLocalizedResourceId(globalElementId)

    if (localizedGlobalElementId == null) return null

    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.LocalizedGlobalElement,
      localizedGlobalElementId,
    )
  }

  async fetchLocalizedGlobalElement(
    globalElementId: string,
  ): Promise<LocalizedGlobalElement | null> {
    const locale = this.locale

    if (locale == null) return null

    // If getLocalizedResourceId returns null, it means we have tried to fetch the resource,
    // but the resource is not available. If we haven't fetched it yet, it'll return undefined.
    const noLocalizedResource = this.getLocalizedResourceId(globalElementId) === null

    if (noLocalizedResource) return null

    // TODO(fikri): We're checking the cache here because the cache hit will fail on fetchAPIResource.
    // This is because in the cache we're saving the ID of the localizedGlobalElementId,
    // but we're reading the cache using the globalElementId.
    // This is very confusing and it can lead to more bugs. We need to refactor how we handle
    // localized global element.
    const cacheResult = this.readLocalizedGlobalElement(globalElementId)

    if (cacheResult) return cacheResult

    const result = await this.makeswiftApiClient.dispatch(
      MakeswiftApiClient.fetchAPIResource(
        APIResourceType.LocalizedGlobalElement,
        globalElementId,
        locale,
      ),
    )

    this.setLocalizedResourceId({
      resourceId: globalElementId,
      localizedResourceId: result?.id ?? null,
    })

    return result
  }

  readPagePathnameSlice(pageId: string): PagePathnameSlice | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.PagePathnameSlice,
      pageId,
    )
  }

  async fetchPagePathnameSlice(pageId: string): Promise<PagePathnameSlice | null> {
    return await this.makeswiftApiClient.dispatch(
      MakeswiftApiClient.fetchAPIResource(APIResourceType.PagePathnameSlice, pageId, this.locale),
    )
  }

  resolvePagePathnameSlice(pageId: string | undefined): FetchableValue<PagePathnameSlice | null> {
    return this.resolveResource(APIResourceType.PagePathnameSlice, {
      id: pageId,
      read: id => this.readPagePathnameSlice(id),
      fetch: id => this.fetchPagePathnameSlice(id),
    })
  }

  resolveResource<R>(
    type: APIResourceType,
    {
      id,
      read,
      fetch,
    }: {
      id: string | undefined
      read: (id: string) => R | null
      fetch: (id: string) => Promise<R | null>
    },
  ): FetchableValue<R | null> {
    const _read = () => (id != null ? read(id) : null)
    let lastValue: R | null = null
    return {
      name: `${type}:${id}`,
      readStable: () => (lastValue = _read()),
      subscribe: (onUpdate: () => void) =>
        this.subscribe(() => {
          if (_read() !== lastValue) onUpdate()
        }),
      fetch: async () => (id != null ? fetch(id) : null),
    }
  }

  readTable(tableId: string): Table | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.Table,
      tableId,
    )
  }

  async fetchTable(tableId: string): Promise<Table | null> {
    return await this.makeswiftApiClient.dispatch(
      MakeswiftApiClient.fetchAPIResource(APIResourceType.Table, tableId),
    )
  }

  async createTableRecord(tableId: string, columns: any): Promise<void> {
    await this.graphqlClient.request<
      CreateTableRecordMutationResult,
      CreateTableRecordMutationVariables
    >(CreateTableRecordMutation, { input: { data: { tableId, columns } } })
  }

  readSite(siteId: string): Site | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.Site,
      siteId,
    )
  }

  readPage(pageId: string): Page | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.Page,
      pageId,
    )
  }

  readSnippet(snippetId: string): Snippet | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.Snippet,
      snippetId,
    )
  }

  private getLocalizedResourceId(resourceId: string): string | undefined | null {
    return this.localizedResourcesMap?.get(resourceId)
  }

  setLocalizedResourceId({
    resourceId,
    localizedResourceId,
  }: {
    resourceId: string
    localizedResourceId: string | null
  }): void {
    this.localizedResourcesMap.set(resourceId, localizedResourceId)
  }
}
