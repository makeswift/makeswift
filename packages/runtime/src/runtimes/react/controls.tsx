import { useRef, ReactNode } from 'react'

import { ControlDefinition } from '@makeswift/controls'

import { partitionRecord } from '../../utils/partition'

import * as ReactPage from '../../state/react-page'
import { isLegacyDescriptor, LegacyDescriptor } from '../../prop-controllers/descriptors'

import { useStore } from './hooks/use-store'

import { resolveLegacyDescriptorProp } from './legacy-controls'
import { useResolvedProps } from './hooks/use-resolved-props'

type PropsValueProps = {
  element: ReactPage.ElementData
  children(props: Record<string, unknown>): ReactNode
}

function useControlDefs(
  elementType: string,
): readonly [Record<string, LegacyDescriptor>, Record<string, ControlDefinition>] {
  const store = useStore()
  const all = ReactPage.getComponentPropControllerDescriptors(store.getState(), elementType) ?? {}
  return useRef(partitionRecord(all, isLegacyDescriptor)).current
}

export function ResolveProps({ element, children: renderComponent }: PropsValueProps): ReactNode {
  const [legacyDescriptors, definitions] = useControlDefs(element.type)

  const resolvedProps = useResolvedProps(definitions, element.props, element.key)

  const renderFn = Object.entries(legacyDescriptors).reduceRight(
    (renderFn, [propName, descriptor]) =>
      props =>
        resolveLegacyDescriptorProp(descriptor, propName, element.props[propName], props, renderFn),
    renderComponent,
  )

  return renderFn(resolvedProps)
}
