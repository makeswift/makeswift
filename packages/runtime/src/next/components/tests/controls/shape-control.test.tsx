/** @jest-environment jsdom */
import { Checkbox, Color, Image, List, Shape } from '@makeswift/controls'
import { testPageControlPropRendering } from './page-control-prop-rendering'
import { APIResourceType } from '../../../../api'

const swatchId = '[swatch-test-id]'
const fileId = '[file-test-id]'

describe('Page', () => {
  test('Shape (heterogenous properties)', async () => {
    await testPageControlPropRendering(
      Shape({
        type: {
          checkbox: Checkbox({ defaultValue: true }),
          color: Color({ defaultValue: 'rgb(255,0,0)' }),
          image: Image({ format: Image.Format.WithDimensions }),
          image2: Image({ format: Image.Format.URL }),
          list: List({
            type: Checkbox({ defaultValue: true }),
          }),
          shape: Shape({
            type: {
              checkbox: Checkbox({ defaultValue: true }),
            },
          }),
        },
      }),
      {
        value: {
          checkbox: true,
          color: { swatchId, alpha: 0.5 },
          image: {
            type: 'makeswift-file',
            id: fileId,
          },
          image2: {
            type: 'external-file',
            url: 'https://example.com/external-jpg.jpg',
            width: 100,
            height: 100,
          },
          list: [true],
          shape: {
            checkbox: true,
          },
        },
        cacheData: {
          apiResources: {
            Swatch: [
              {
                id: swatchId,
                value: {
                  __typename: APIResourceType.Swatch,
                  id: swatchId,
                  hue: 238,
                  saturation: 87,
                  lightness: 49,
                },
              },
            ],
            File: [
              {
                id: fileId,
                value: {
                  __typename: APIResourceType.File,
                  id: fileId,
                  publicUrl: 'https://example.com/image.png',
                  name: 'image',
                  extension: 'png',
                  dimensions: {
                    width: 100,
                    height: 100,
                  },
                },
              },
            ],
          },
        },
        expectedRenders: 1,
      },
    )
  })
})
