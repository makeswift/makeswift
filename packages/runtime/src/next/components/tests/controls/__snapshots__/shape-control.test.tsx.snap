// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page Shape (heterogenous properties): resolvedValue 1`] = `
{
  "checkbox": true,
  "color": "rgba(16, 24, 234, 0.5)",
  "image": {
    "dimensions": {
      "height": 100,
      "width": 100,
    },
    "url": "https://example.com/image.png",
  },
  "image2": "https://example.com/external-jpg.jpg",
  "list": [
    true,
  ],
  "shape": {
    "checkbox": true,
  },
}
`;

exports[`Page Shape (heterogenous properties): snapshot 1`] = `
{
  "cacheData": {
    "apiResources": {
      "File": [
        {
          "id": "[file-test-id]",
          "value": {
            "__typename": "File",
            "dimensions": {
              "height": 100,
              "width": 100,
            },
            "extension": "png",
            "id": "[file-test-id]",
            "name": "image",
            "publicUrl": "https://example.com/image.png",
          },
        },
      ],
      "Swatch": [
        {
          "id": "[swatch-test-id]",
          "value": {
            "__typename": "Swatch",
            "hue": 238,
            "id": "[swatch-test-id]",
            "lightness": 49,
            "saturation": 87,
          },
        },
      ],
    },
    "localizedResourcesMap": {},
  },
  "document": {
    "data": {
      "key": "00000000-0000-0000-0000-000000000000",
      "props": {
        "children": {
          "columns": [
            {
              "deviceId": "desktop",
              "value": {
                "count": 12,
                "spans": [
                  [
                    12,
                  ],
                ],
              },
            },
          ],
          "elements": [
            {
              "key": "11111111-1111-1111-1111-111111111111",
              "props": {
                "propKey": {
                  "checkbox": {
                    "@@makeswift/type": "checkbox::v1",
                    "value": true,
                  },
                  "color": {
                    "@@makeswift/type": "color::v1",
                    "alpha": 0.5,
                    "swatchId": "[swatch-test-id]",
                  },
                  "image": {
                    "id": "[file-test-id]",
                    "type": "makeswift-file",
                    "version": 1,
                  },
                  "image2": {
                    "height": 100,
                    "type": "external-file",
                    "url": "https://example.com/external-jpg.jpg",
                    "version": 1,
                    "width": 100,
                  },
                  "list": [
                    {
                      "id": "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx",
                      "type": "makeswift::controls::checkbox",
                      "value": {
                        "@@makeswift/type": "checkbox::v1",
                        "value": true,
                      },
                    },
                  ],
                  "shape": {
                    "checkbox": {
                      "@@makeswift/type": "checkbox::v1",
                      "value": true,
                    },
                  },
                },
              },
              "type": "TestComponent",
            },
          ],
        },
      },
      "type": "./components/Root/index.js",
    },
    "fonts": [],
    "id": "test-page-id",
    "locale": null,
    "localizedPages": [],
    "meta": {},
    "seo": {},
    "site": {
      "id": "test-site-id",
    },
    "snippets": [],
  },
}
`;
